### ISSUE TITLE
Testing Crabler

### ISSUE THREAD

**Issue by Susurrus** [Gonzih/crabler#12]
As a follow-up to !9 I wanted to implement some testing in Crabler. I wanted to do some tests around the response status and also URL handling with the 301 & 302 redirects. There are two things that need to happen to make this work: one is a mocking library that can respond to HTTP requests and the other is a mocking library to mock the callbacks for `on_response` and `on_html`. 

[Mockito](https://docs.rs/mockito/latest/mockito/) seems a reasonable choice of HTTP mocking library. And it seems straightforward to implement. However, what I've gotten stuck on is how to mock the callbacks. I've been looking into [mockall](https://docs.rs/crate/mockall/0.11.0) (and there are some others like it like [mockiato](https://docs.rs/mockiato/0.9.5/mockiato/)), but these want to mock the entire trait and then you need to use a Double or the mock object directly. However I don't see a straightforward way to do this.

I'd love to help implement more testing for crabler, but right now I don't know how to do that with its current architecture.

**Comment by Susurrus** [comment]
[Mocktupus](https://docs.rs/mocktopus/latest/mocktopus/index.html) looks like an option that might be feasible. Testing it with code like:

```rust
use crabler::*;
use mockito::{mock, server_url};
use mocktopus::macros::*;

#[derive(WebScraper)]
#[on_response(response_handler)]
#[on_html("a[href]", link_handler)]
struct Scraper {}

#[mockable]
impl Scraper {
    async fn response_handler(&mut self, response: Response) -> Result<()> {
        Ok(())
    }

    async fn link_handler(&mut self, _response: Response, a: Element) -> Result<()> {
        Ok(())
    }
}

#[async_std::test]
async fn test_200() {
    use mocktopus::mocking::*;

    Scraper::response_handler.mock_safe(|_, _| MockResult::Return(Ok(())));
    Scraper::link_handler.mock_safe(|_, _, _| MockResult::Return(Ok(())));

    let _m = mock("GET", "/")
      .with_status(201)
      .with_header("content-type", "text/html")
      .with_body("<html><body><a href=\"https://rust-lang.org\">Rust</a></body></html>")
      .create();

    let scraper = Scraper {};
    scraper.run(Opts::new().with_urls(vec![&server_url()]).with_threads(1)).await;
}
```

But I get errors about `crabler::Element` not implementing `Send` and the return values I'm injecting seem to cause errors also. I'll need to play with this more.

**Comment by Gonzih** [comment]
One way would be to abstract url fetching from WebScraper into a separate trait (lets call it HttpTransport for example). This way default would use async http client and if you need testing you can implement your own transport that would just do hashmap lookups.

**Comment by Susurrus** [comment]
I started looking into this and was trying to find if anyone else had already solved this problem, as I imagine someone has! I found that `tide` went the path of running their own HTTP server and querying that with generated data rather than trying to mock ([see here](https://github.com/http-rs/tide/blob/d81eabc2cb775cb1750584502ac90e049ed9e34f/tests/server.rs#L10)). Seems pretty heavy-handed, but I guess that'll be the Plan B here.

**Comment by Susurrus** [comment]
I think the right approach is to instead of doing one-off `surf::get()` calls, create and keep a `surf::client` instance and use a custom middleware to embed fake responses.

Note that the problem I'm looking to solve here isn't testing of my downstream application, but for this library itself. I've run into some panics coming from the redirection logic that I implemented and before I rely too much on this library for real work I want to ensure that it's working as expected for the HTTP responses I expect to see. Fuzz testing would likely be something I'd want to further extend here to make sure this puppy never crashes!

**Comment by Gonzih** [comment]
keeping instance of surf::client also works, give it a try, it sounds like a good way forward!

**Comment by Gonzih** [comment]
Maybe it can be done via Opts? `Opts::with_surf_client(instance)` or something like that?

**Comment by Susurrus** [comment]
Using the client approach seems like a good path forward because there's actually a [built-in middleware](https://docs.rs/surf/latest/surf/middleware/struct.Redirect.html) for following redirects, replacing all of my custom code! So my first stab at this is going to be to refactor the existing code to use this redirect middleware. I think there will be minimal performance impact from `clone`ing the client into each `Worker` as I think `Worker`s are somewhat long lived. There is an additional request for each query when using their middleware, but since people can turn this off, I think it's reasonable.

**Comment by Gonzih** [comment]
yeah clone overhead is fine, i think currently there is overhead of surf creating new client for each request (maybe im mistaken on this) under the hood.

**Comment by Susurrus** [comment]
So the main issue here isn't actually faking the HTTP requests, as `mockito` makes that trivial. The biggest issue is actually mocking out the callbacks fed into `on_response()` and `on_html`. I'm wondering if we could add the mock infra into the WebScraper for when `cfg(test)` and then you can access those directly during tests. That would also allow downstream apps to do tests, and it wouldn't expose them during normal builds.

**Comment by Susurrus** [comment]
I don't know how to architect that yet, but probably my next step is to see the generated Rust from the macro to see what the complete code is and then go from there. I'm pretty certain there's a way to do that, but I'll have to look it up.

**Comment by Gonzih** [comment]
yes, take a look at derive crate in this repo, it was heavily inspired by rocket.rs

**Comment by Susurrus** [comment]
```
$ cargo install cargo-expand
$ cargo expand --test crawl
```

Gave me what I needed. So I'm thinking the way to do this is to get a mocked function that is called. Let's see where we stand here:
 * [mockiato](https://github.com/mockiato/mockiato) doesn't compile on the latest nightly or stable.
 * [mockall](https://docs.rs/mockall/latest/mockall/) wants to override the entire structs, so that wouldn't let us just check the callback calls.
 * [mock-it](https://github.com/nathanielsimard/mock-it/) should do what we want, where we implement another field on the struct and can call that and then check what those calls involved later, but it requires `CrablerError` to be `Clone` (can't because `std::io::Error` isn't) and for `crabler::Response` to be `PartialEq`, which we could probably manually implement if we ignore the `workinput_tx` and `counter` fields.

I'm thinking now that creating some internal state of `Vec`s that can store tuples of the input args to the functions that can also be validated afterwards could work. Would require `WebScraper::run` to not consume the struct. There are already issues with this implementation however, as `http-types::Response` when cloned clears the body ([link](https://docs.rs/http-types/2.12.0/http_types/struct.Response.html#impl-Clone)).
